epochStart/shardchain/validatorInfoProcessor.go

```diff
@@ -164,58 +163,70 @@ func (vip *ValidatorInfoProcessor) processAllPeerMiniBlocks(metaBlock *block.Met
 }
 
 func (vip *ValidatorInfoProcessor) computeMissingPeerBlocks(metaBlock *block.MetaBlock) {
-	missingNumber := uint32(0)
-	vip.mutMiniBlocksForBlock.Lock()
+	numMissingPeerMiniblocks := uint32(0)
+	vip.mutAllMissingPeerMiniblocks.Lock()
 
 	for _, mb := range metaBlock.MiniBlockHeaders {
 		if mb.Type != block.PeerBlock {
 			continue
 		}
 
-		vip.allPeerMiniblocks[string(mb.Hash)] = &miniBlockInfo{}
+		vip.mapAllMisingPeerMiniblocks[string(mb.Hash)] = nil
 
 		mbObjectFound, ok := vip.miniBlocksPool.Peek(mb.Hash)
 		if !ok {
-			missingNumber++
+			numMissingPeerMiniblocks++
 			continue
 		}
 
 		mbFound, ok := mbObjectFound.(*block.MiniBlock)
 		if !ok {
-			missingNumber++
+			numMissingPeerMiniblocks++
 			continue
 		}
 
-		vip.allPeerMiniblocks[string(mb.Hash)] = &miniBlockInfo{mb: mbFound}
+		vip.mapAllMisingPeerMiniblocks[string(mb.Hash)] = mbFound
 	}
 
-	vip.numMissing = missingNumber
-	vip.mutMiniBlocksForBlock.Unlock()
+	vip.numMissingPeerMiniblocks = numMissingPeerMiniblocks
+	vip.mutAllMissingPeerMiniblocks.Unlock()
 }
 
-func (vip *ValidatorInfoProcessor) retrieveMissingBlocks() error {
-	vip.mutMiniBlocksForBlock.Lock()
+func (vip *ValidatorInfoProcessor) retrieveMissingBlocks() (map[string]*block.MiniBlock, error) {
+	vip.mutAllMissingPeerMiniblocks.Lock()
 	missingMiniblocks := make([][]byte, 0)
-	for mbHash, mbInfo := range vip.allPeerMiniblocks {
-		if mbInfo.mb == nil {
+	for mbHash, mb := range vip.mapAllMisingPeerMiniblocks {
+		if mb == nil {
 			missingMiniblocks = append(missingMiniblocks, []byte(mbHash))
 		}
 	}
-	vip.numMissing = uint32(len(missingMiniblocks))
-	vip.mutMiniBlocksForBlock.Unlock()
+	vip.numMissingPeerMiniblocks = uint32(len(missingMiniblocks))
+	vip.mutAllMissingPeerMiniblocks.Unlock()
 
 	if len(missingMiniblocks) == 0 {
-		return nil
+		return nil, nil
 	}
 
 	go vip.requestHandler.RequestMiniBlocks(core.MetachainShardId, missingMiniblocks)
 
 	select {
 	case <-vip.chRcvAllMiniblocks:
-		return nil
-	case <-time.After(time.Second):
-		return process.ErrTimeIsOut
+		return nil, nil
+	case <-time.After(waitTime):
+		return vip.getCopyOfAllMissingPeerMiniblocks(), process.ErrTimeIsOut
+	}
+}
+
+func (vip *ValidatorInfoProcessor) getCopyOfAllMissingPeerMiniblocks() map[string]*block.MiniBlock {
```

Comment:
if these are missing then you have a nil pointer for every missing Peer miniblock ?

Reply:
here you are effectively returning all the miniblocks, those found and those missing as well. need a better naming.

Reply:
Refactored